lib/tower-http-digest/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+use std::{
+    error::Error as StdError,
+    future::Future,
+    pin::Pin,
+    str::FromStr,
+    task::{self, ready, Poll},
+};
+use strum::{AsRefStr, EnumString};
+use tower_layer::Layer;
+use tower_service::Service;
+
+type BoxError = Box<dyn StdError + Send + Sync>;
+
+static DIGEST_HEADER_NAME: HeaderName = HeaderName::from_static("digest");
+
+#[derive(AsRefStr, Clone, Copy, Default, EnumString)]
+#[non_exhaustive]
+pub enum Algorithm {
+    #[default]
+    #[strum(ascii_case_insensitive, serialize = "sha-256")]
+    Sha256,
+
+    #[strum(ascii_case_insensitive, serialize = "sha-512")]
+    Sha512,
+}
+
+impl Algorithm {
+    pub fn digest(&self, data: impl AsRef<[u8]>) -> Vec<u8> {
+        match self {
+            Self::Sha256 => Sha256::digest(data).to_vec(),
+            Self::Sha512 => Sha512::digest(data).to_vec(),
+        }
+    }
+}
+
+pin_project! {
+    #[project = DigestFutureProj]
+    pub enum DigestFuture<S, F> {
+        BuildingDigest {
+            service: S,
+
+            algorithm: Algorithm,
+            parts: Option<http::request::Parts>,
+
+            #[pin]
+            body: Body,
+            body_accumulator: Option<BytesMut>,
+        },
+        PollServiceFuture {
+            #[pin]
+            future: F,
+        },
+    }
+}
+
+impl<S> Future for DigestFuture<S, S::Future>
+where
+    S: Service<Request<Body>>,
+    S::Error: Into<BoxError>,
+{
+    type Output = Result<S::Response, BoxError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        loop {
+            match self.as_mut().project() {
+                DigestFutureProj::BuildingDigest {
+                    service,
+                    algorithm,
+                    parts,
+                    mut body,
+                    body_accumulator,
+                } => {
+                    while let Some(frame) = ready!(body.as_mut().poll_frame(cx))
+                        .transpose()
+                        .map_err(BoxError::from)?
+                    {
+                        if let Ok(data) = frame.into_data() {
+                            let accumulator = body_accumulator
+                                .as_mut()
+                                .expect("[Bug] Missing accumulator");
+
+                            accumulator.put(data);
+                        }
+                    }
+
+                    let digest_header = parts.as_ref().unwrap().headers.get(&DIGEST_HEADER_NAME);
+                    let algorithm = if let Some(digest_header) = digest_header {
+                        let Some((algorithm_name, ..)) = digest_header.to_str()?.split_once('=')
+                        else {
+                            return Poll::Ready(Err("Invalid header value".into()));
+                        };
+                        Algorithm::from_str(algorithm_name)?
+                    } else {
+                        *algorithm
+                    };
```

Comment:
I guess we could dispatch that with an `Either` future, since that would result in two separate futures in the error case (or we add another case to the state machine)

Reply:
Yep. Implemented that. It's now a separate state the state machine transitions through.  

`ParseHeader` -> `BuildingDigest` -> `PollServiceFuture`

Reply:
I just did it this way since it's not much more complicated than doing it in the `call` impl, and I was able to keep the `call` impl as clean as it is right now.